programs/bpf_loader/src/syscalls.rs

```diff
@@ -2540,6 +2573,137 @@ fn call<'a>(
     Ok(SUCCESS)
 }
 
+// Return data handling
+pub struct SyscallSetReturnData<'a> {
+    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallSetReturnData<'a> {
+    fn call(
+        &mut self,
+        addr: u64,
+        len: u64,
+        _arg3: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        let mut invoke_context = question_mark!(
+            self.invoke_context
+                .try_borrow_mut()
+                .map_err(|_| SyscallError::InvokeContextBorrowFailed),
+            result
+        );
+
+        let budget = invoke_context.get_compute_budget();
+
+        question_mark!(
+            invoke_context
+                .get_compute_meter()
+                .consume(len / budget.cpi_bytes_per_unit + budget.syscall_base_cost),
+            result
+        );
+
+        if len > MAX_RETURN_DATA as u64 {
+            *result = Err(SyscallError::ReturnDataTooLarge(len, MAX_RETURN_DATA as u64).into());
+            return;
+        }
+
+        let return_data = question_mark!(
+            translate_slice::<u8>(memory_mapping, addr, len, self.loader_id, true),
+            result
+        );
+
+        let program_id = *question_mark!(
+            invoke_context
+                .get_caller()
+                .map_err(SyscallError::InstructionError),
+            result
+        );
+
+        invoke_context.set_return_data(Some((program_id, return_data.to_vec())));
+
+        *result = Ok(0);
+    }
+}
+
+pub struct SyscallGetReturnData<'a> {
+    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallGetReturnData<'a> {
+    fn call(
+        &mut self,
+        return_data_addr: u64,
+        len: u64,
+        program_id_addr: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        let invoke_context = question_mark!(
+            self.invoke_context
+                .try_borrow()
+                .map_err(|_| SyscallError::InvokeContextBorrowFailed),
+            result
+        );
+
+        let budget = invoke_context.get_compute_budget();
+
+        if let Some((program_id, return_data)) = invoke_context.get_return_data() {
```

Comment:
Why not return 0 if `None`?

Reply:
The idea is that a program can set an empty return data.

Having said that, I'm still not sure about this. This is overloading the length return with  a special "nothing set" value.

Reply:
Have a use case in mind where a program would set an empty return value?/

Reply:
No, I don't I must admit. The only thought I had was that with program_id recorded, we need an option anyway so now empty return data is easier to implement.

For solidity only have functions with no return value will return 0 bytes, in which case we wouldn't be looking for return data anyway.

What do you think?

Reply:
Yeah, I'm thinking if len is zero we return zero and don't touch the output data or program id

Reply:
I've changed it. If a return data of zero length is set, then the return_data Option becomes none and this function return zero.